UDP hole punching MAID-1428 and MAID-1429

canndrew opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/maidsafe/crust/363)

<!-- Reviewable:end -->

inetic committed:
Adds SequenceNumber struct

To deal with sequence wrap arounds

inetic committed:
Adds util function to check if an IP is on the same subnet

inetic committed:
Adds heuristic_geo_cmp function

To try to determine which IP address is closer to us

canndrew committed:
Add client functions for udp hole punching RFC

inetic committed:
Merge remote-tracking branch 'andrew/udp_hole_punching' into MAID-1429

inetic committed:
Use match wildcard '_' for events in tests and examples

canndrew committed:
udp hole punching RFC server-side stuff

inetic committed:
Adds SequenceNumber object to the State struct

inetic committed:
Removes warnings

inetic committed:
Makes use of hole_punching::blocking_get_mapped_udp_socket

In Service and State

inetic committed:
WIP [skip ci]

canndrew committed:
Try utp hole punching servers one-by-one instead of all simultaneously

canndrew committed:
Add hole punching test. Fix WrapSocketAddr encode/decode

inetic committed:
Merge branch 'MAID-1429' into udp_hole_punching

Conflicts:
	src/service.rs

inetic committed:
Merge branch 'master' of github.com:maidsafe/crust into udp_hole_punching

inetic committed:
Moves udp hole punching logic into hole_punching module

inetic committed:
Adds test to check if hole punching terminates

It currently does not

inetic committed:
Adds more udp hole punching tests

inetic committed:
Adds asertion to not exceed maximum datagram size

Seems cbor doesn't serialize into a fixed size for some reason

inetic committed:
Increment MAX_DATAGRAM_SIZE from 16 to 128

To be safe, for some reason Cbor doesn't serialize into
bytes of exact length

inetic committed:
More precise assertions in hole punch tests

inetic committed:
Udp hole punch returns TimeOut on time out

Not WouldBlock

inetic committed:
Refactores udp hole punch tests a bit

inetic committed:
Replaces dummy mapper port with the one from hole puhch server

canndrew committed:
Refactoring and bugfixes

maidsafe-highfive commented:
Thanks for the pull request, and welcome! The MaidSafe team is excited to review your changes, and you should hear from @inetic (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTOR.md](../blob/master/CONTRIBUTOR) for more information.

inetic committed:
blocking_udp_punch_hole now takes only one peer SocketAddr

Sending to two or more addresses would fail to make hole
if we're behind an Address or Port Restricted Code NAT.

inetic committed:
Merge remote-tracking branch 'ms/master' into udp_hole_punching

inetic committed:
Separates parsing from interpreting in crust_peer example

inetic committed:
Impl Encodable/Decodable for util::SocketAddrW

And exports SocketAddrW

inetic committed:
Moves PeerEndpoint struct to another place in the crust_peer.rs file

inetic committed:
Only wait for bootstrap if running speed test

For one, it is not desirable otherwise, and the waiting
code is broken and I'm currently working on something else.

inetic committed:
Adds hole punching code to crust_peer

inetic committed:
Fixes warnings (errors due to lint)

inetic committed:
In hole punch tests, use loopback instead of 0.0.0.0

Hope to fix tests on windows

canndrew committed:
use ::time::SteadyTime instead of ::time::Tm

canndrew committed:
Reset destination in blocking_udp_punch_hole
